public/js/selfoss-db.js

```diff
@@ -1,81 +1,226 @@
+                        selfoss.dbOnline.syncing.reject();
+                        stopChecking = true;
+                    }
+                } else {
+                    stopChecking = true;
+                }
+
+                if (stopChecking) {
+                    window.clearInterval(monitor);
+                }
+            }, 10000);
+        }
+
+        return selfoss.dbOnline.syncing;
+    },
+
+
+    _syncDone: function(success) {
+        success = (typeof success !== 'undefined') ? success : true;
+
+        if (selfoss.dbOnline.syncing) {
+            if (success) {
+                selfoss.dbOnline.syncing.resolve();
+            } else {
+                var request = selfoss.dbOnline.syncing.request;
+                selfoss.dbOnline.syncing.reject();
+                if (request) {
+                    request.abort();
+                }
+            }
+        }
+    },
+
+
     /**
      * sync server status.
      *
-     * @return void
+     * @return Promise
      */
-    sync: function(force) {
-        force = (typeof force !== 'undefined') ? force : false;
-
-        if (!force && (selfoss.lastUpdate == null ||
-                       Date.now() - selfoss.lastSync < 5 * 60 * 1000)) {
-            var d = $.Deferred();
-            d.resolve();
-            return d; // ensure any chained function runs
+    sync: function(updatedStatuses, chained) {
+        if (selfoss.dbOnline.syncing && !chained) {
+            if (updatedStatuses) {
+                // Ensure the status queue is not cleared and gets sync'ed at
+                // next sync.
+                var d = $.Deferred();
+                d.reject();
+                return d;
+            } else {
+                return selfoss.dbOnline.syncing;
+            }
         }
 
+        var syncing = selfoss.dbOnline._syncBegin();
+
         var getStatuses = true;
-        if (selfoss.lastUpdate == null) {
-            selfoss.lastUpdate = new Date(0);
+        if (selfoss.db.lastUpdate === null) {
+            selfoss.db.lastUpdate = new Date(0);
             getStatuses = undefined;
         }
 
-        return $.ajax({
+        var syncParams = {
+            since: selfoss.db.lastUpdate.toISOString(),
+            tags: true,
+            sources: selfoss.filter.sourcesNav ? true : undefined,
+            itemsStatuses: getStatuses
+        };
+
+        if (updatedStatuses && updatedStatuses.length > 0) {
+            syncParams.updatedStatuses = updatedStatuses;
+        }
+
+        if (selfoss.db.storage) {
+            syncParams.itemsSinceId = selfoss.dbOffline.lastItemId;
+            syncParams.itemsNotBefore = selfoss.dbOffline.newestGCedEntry.toISOString();
+            syncParams.itemsHowMany = selfoss.filter.itemsPerPage;
+        }
+
+        selfoss.dbOnline.statsDirty = false;
+
+        syncing.request = $.ajax({
             url: 'items/sync',
-            type: 'GET',
+            type: updatedStatuses ? 'POST' : 'GET',
```

Comment:
Why not always use `POST`?

Reply:
```POST``` changes things whereas ```GET``` does not change the database. This is a ```HTTP``` convention.

Reply:
I guess I view `POST /items/sync` as an endpoint that changes `x` things, where `x` can be zero as well. I guess, in true REST, the endpoint would be split into two, one for each synchronization direction.

Reply:
I chose this because in one request, changes are committed to the server and new items are retrieved. On a slow link, this may count.